pkg/terraform/terraform.go

```diff
@@ -111,13 +113,143 @@ type TerraformEntrypoint struct {
+	}
+	entrypointToModules := make(map[string]map[string]struct{})
+	modulesToEntrypoints := make(map[string]map[string]struct{})
+	for _, entrypoint := range entrypoints {
+		entrypointToModules[entrypoint.Path] = make(map[string]struct{})
+		recurseAndAppend(entrypoint.Path, entrypoint.Path, entrypointToModules, moduleUsages)
+	}
+	for _, modules := range entrypointToModules {
+		for module := range modules {
+			if _, ok := modulesToEntrypoints[module]; !ok {
+				modulesToEntrypoints[module] = make(map[string]struct{})
+			}
+		}
+	}
+	for module := range modulesToEntrypoints {
+		for entrypoint, modules := range entrypointToModules {
+			if _, modOK := modules[module]; modOK {
+				modulesToEntrypoints[module][entrypoint] = struct{}{}
+			}
+		}
+	}
+	return &ModuleUsageGraph{
+		EntrypointToModules:  entrypointToModules,
+		ModulesToEntrypoints: modulesToEntrypoints,
+	}, nil
+}
+
+func recurseAndAppend(rootPth, pth string, entrypointToModules map[string]map[string]struct{}, moduleUsages map[string]*Modules) {
+	if usage, ok := moduleUsages[pth]; ok {
+		for modulePath := range usage.ModulePaths {
+			entrypointToModules[rootPth][modulePath] = struct{}{}
+			recurseAndAppend(rootPth, modulePath, entrypointToModules, moduleUsages)
+		}
+	}
+}
+
+// Modules represents the details of the modules used by a given module or terraform entrypoint.
+type Modules struct {
+	ModulePaths        map[string]struct{}
+	ModuleOrEntrypoint string
+}
+
+// modules locates all terraform entrypoints or modules and finds all of their module usages.
+func modules(ctx context.Context, rootDir string, skipUnresolvableModules bool) (map[string]*Modules, error) {
+	logger := logging.FromContext(ctx)
+	matches := make(map[string]*Modules)
+	if err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk directory %s: %w", path, err)
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		if filepath.Ext(path) != ".tf" {
+			return nil
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s: %w", path, err)
+		}
+
+		// Read the contents of the file
+		contents, err := io.ReadAll(file)
+		if err != nil {
+			return fmt.Errorf("failed to read contents of file %s: %w", path, err)
+		}
+		defer file.Close()
+
+		m := moduleSourcePattern.FindAllStringSubmatch(string(contents), -1)
+		if len(m) == 0 {
+			return nil
+		}
+
+		absPath, err := util.PathEvalAbs(path)
+		if err != nil {
+			return fmt.Errorf("failed to get absolute path for directory %s: %w", rootDir, err)
+		}
+
+		if _, ok := matches[filepath.Dir(absPath)]; !ok {
+			matches[filepath.Dir(absPath)] = &Modules{
+				ModuleOrEntrypoint: filepath.Dir(absPath),
+				ModulePaths:        make(map[string]struct{}),
+			}
+		}
+		modulePaths := matches[filepath.Dir(absPath)].ModulePaths
+
+		for _, ms := range m {
+			if len(ms) == 2 {
```

Comment:
Might be worth inverting this match to save a level of indentation.

```
if len(ms) != 2 {
  continue
}
relativeModulePath := ...
```

Reply:
Done